sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpMessageConverter.cs

```diff
@@ -540,24 +546,34 @@ public virtual ServiceBusReceivedMessage AmqpMessageToSBMessage(AmqpMessage amqp
                 }
             }
 
+            return annotatedMessage;
+        }
+
+        public virtual ServiceBusReceivedMessage AmqpMessageToSBReceivedMessage(AmqpMessage amqpMessage, bool isPeeked = false)
+        {
+            AmqpAnnotatedMessage annotatedMessage = AmqpMessageToAnnotatedMessage(amqpMessage, isPeeked);
+
+            ServiceBusReceivedMessage sbMessage = new ServiceBusReceivedMessage(annotatedMessage);
+
             // lock token
 
-            if (amqpMessage.DeliveryTag.Count == GuidSizeInBytes)
-            {
-                Span<byte> guidBytes = stackalloc byte[GuidSizeInBytes];
-                amqpMessage.DeliveryTag.AsSpan().CopyTo(guidBytes);
-                if (!MemoryMarshal.TryRead<Guid>(guidBytes, out var lockTokenGuid))
-                {
-                    lockTokenGuid = new Guid(guidBytes.ToArray());
-                }
-                sbMessage.LockTokenGuid = lockTokenGuid;
-            }
+            sbMessage.LockTokenGuid = ParseGuidBytes(amqpMessage.DeliveryTag);
 
             amqpMessage.Dispose();
 
             return sbMessage;
         }
 
+        public virtual Guid ParseGuidBytes(ReadOnlyMemory<byte> bytes)
+        {
+            if (bytes.Length == GuidSizeInBytes)
+            {
+                return new Guid(bytes.Span.ToArray());
```

Comment:
The previous code was optimized to not do the heap allocation in the majority of the cases now it does it again all the time. Why was this changed? 

Essentially I tried to by compliant here with .NET Standard 2.0 as well as keeping the copying mechanics in place when I optimized this code before. Unfortunately, we do not have the [GUID constructor available](https://learn.microsoft.com/en-us/dotnet/api/system.guid.-ctor?view=net-7.0#system-guid-ctor(system-readonlyspan((system-byte)))) that allows passing in the span in .NET Standard 2.0 but in theory the previous code could be optimized to `#if def` the `MemoryMarshal` part against a specific TFM that supports that constructor. 

But I would definitely try to preserve the previous memory characteristics of this code unless we can see that the before and after is not worse, unless you are hoping for me to optimize it again ;)

Reply:
Damn.   I never remember to check whether or not the `Span` overload is available for `netstandard2.0` - that explains the pattern that was there.

Reply:
@KrzysztofCwalina mentioned that TryRead allocates anyway - https://github.com/Azure/azure-sdk-for-net/pull/33682#discussion_r1088077365

Reply:
If we want to avoid this allocation in NS2.0, we could do:
```csharp
Int32 a = BinaryPrimitives.ReadInt32LittleEndian(span);
Int16 b = BinaryPrimitives.ReadInt16LittleEndian(span.Slice(sizeof(Int32)));
Int16 c = BinaryPrimitives.ReadInt16LittleEndian(span.Slice(sizeof(Int32) + sizeof(Int16)));
var result = new Guid(a, b, c, span[8], span[9], span[10], span[11], span[12], span[13], span[14], span[15]);
return result;
```

Reply:
in addition we would need to have another code path of big endian: https://github.com/dotnet/runtime/blob/main/src/libraries/System.Private.CoreLib/src/System/Guid.cs#L49

Reply:
And BTW, I did not say that the previous code allocates. I said that it copies the data :-)

Reply:
When I originally contributed the version that was changed here we discussed the copying and decided keep the copying semantics. All I did was contributing a better version than the original that was faster and less allocation heavy. Didn't go down the path of a non-copying version. I still have the original benchmark available and can easily compare it to the current if there is interest

Reply:
So essentially the TryRead appears to copy without allocating, whereas when we call the Guid constructor we need to allocate due to the NS2.0 API. But if TryRead copies, why the need for the stackalloc span?

Reply:
I extended the previous benchmark I used to compare the memory marshal version against the original buffer block copy version

```csharp
using System.Runtime.CompilerServices;

namespace MicroBenchmarks.ServiceBus
{
    using System;
    using System.Runtime.InteropServices;
    using BenchmarkDotNet.Attributes;
    using BenchmarkDotNet.Configs;
    using BenchmarkDotNet.Diagnosers;
    using BenchmarkDotNet.Exporters;
    using BenchmarkDotNet.Jobs;

    [Config(typeof(Config))]
    public class GuidConvertion
    {
        private ArraySegment<byte> data;
        private ReadOnlyMemory<byte> dataAsMemory;

        private class Config : ManualConfig
        {
            public Config()
            {
                AddExporter(MarkdownExporter.GitHub);
                AddDiagnoser(MemoryDiagnoser.Default);
                AddJob(Job.Default.WithInvocationCount(15600000));
            }
        }
        [IterationSetup]
        public void Setup()
        {
            data = new ArraySegment<byte>(Guid.NewGuid().ToByteArray());
            dataAsMemory = new ReadOnlyMemory<byte>(data.Array, data.Offset, data.Count);
        }

        [Benchmark]
        public Guid BufferAndBlockCopy()
        {
            var guidBuffer = new byte[16];
            Buffer.BlockCopy(data.Array, data.Offset, guidBuffer, 0, 16);
            return new Guid(guidBuffer);
        }

        [Benchmark()]
        public Guid SpanAndMemoryMarshal()
        {
            Span<byte> guidBytes = stackalloc byte[16];
            data.AsSpan().CopyTo(guidBytes);
            if (!MemoryMarshal.TryRead<Guid>(guidBytes, out var lockTokenGuid))
            {
                lockTokenGuid = new Guid(guidBytes.ToArray());
            }
            return lockTokenGuid;
        }
        
        [Benchmark()]
        [SkipLocalsInit]
        public Guid SpanAndMemoryMarshalLocalsInit()
        {
            Span<byte> guidBytes = stackalloc byte[16];
            data.AsSpan().CopyTo(guidBytes);
            if (!MemoryMarshal.TryRead<Guid>(guidBytes, out var lockTokenGuid))
            {
                lockTokenGuid = new Guid(guidBytes.ToArray());
            }
            return lockTokenGuid;
        }
        
        [Benchmark(Baseline = true)]
        public Guid SpanToArray()
        {
            return new Guid(dataAsMemory.Span.ToArray());;
        }
    }
}
```

![image](https://user-images.githubusercontent.com/174258/215579064-70c91204-6222-4c7b-88e5-0ed57aff9a7b.png)

Reply:
@danielmarbach can you clarify why we need the stackalloc span, as opposed to just using TryRead with data.AsSpan?

Reply:
Let me reiterate what I mentioned before (probably worded really poorly). The original PR is this one

https://github.com/Azure/azure-sdk-for-net/pull/19857

In that my intent was to improve the existing version that did take a defensive copy. I assumed that the defensive copy was necessary due to the underlying buffer management of the Amqp Library. To keep the surface area of change small, I went ahead and replaced the block copy version with a more efficient version _that still creates a defensive_ copy. 

Whether those assumptions were correct I don't know, but it was approved by two awesome people ;)

By looking at the conversation started by @jsquire in https://github.com/Azure/azure-sdk-for-net/pull/19857#discussion_r602297286 I guess we have never settled further on the "defensive copy" part of the discussion. So in case the defensive copy is not needed, then you could definitely `TryRead` directly.

Reply:
And yeah that would be :speedboat: 

![image](https://user-images.githubusercontent.com/174258/215589295-3c2eb46d-1c1f-496b-9ebb-3daac270dc4e.png)

Reply:
Got it - thanks for the pointers. I think the point is that the defensive copy with a stackalloc span is unnecessary - if TryRead didn't copy then it would be a bug because the span would go away; if TryRead does copy (which is what happens) then it is unnecessary.

Reply:
Looking at that comment, at least I'm consistent in never remembering to look at which netstandard versions support the Span overload.   

<img src="https://media0.giphy.com/media/HfFccPJv7a9k4/giphy.gif"/>

Reply:
Will update to add back the TryRead.

Reply:
Seen now @KrzysztofCwalina great remark in https://github.com/Azure/azure-sdk-for-net/pull/33682#discussion_r1088077365 and yeah I haven't considered that aspect. One never stops learning :sunglasses: